CommonTools/ParticleFlow/plugins/PFCandidateRecalibrator.cc

```diff
@@ -0,0 +1,283 @@
+//    - a ValueMap<reco::PFCandidateRef> that maps the old to the new, and vice-versa
+
+#include "FWCore/Framework/interface/Frameworkfwd.h"
+#include "FWCore/Framework/interface/EDProducer.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/Framework/interface/ESHandle.h"
+
+#include "FWCore/Framework/interface/Event.h"
+
+#include "CalibFormats/HcalObjects/interface/HcalDbService.h"
+#include "CalibFormats/HcalObjects/interface/HcalDbRecord.h"
+#include "CalibFormats/HcalObjects/interface/HcalCalibrations.h"
+#include "Geometry/CaloTopology/interface/HcalTopology.h"
+#include "Geometry/HcalTowerAlgo/interface/HcalGeometry.h"
+#include "Geometry/CaloGeometry/interface/CaloSubdetectorGeometry.h"
+#include "Geometry/CaloGeometry/interface/CaloGeometry.h"
+#include "Geometry/Records/interface/CaloGeometryRecord.h"
+
+#include "DataFormats/ParticleFlowCandidate/interface/PFCandidateFwd.h"
+#include "DataFormats/ParticleFlowCandidate/interface/PFCandidate.h"
+
+#include "DataFormats/Common/interface/View.h"
+#include "DataFormats/Common/interface/Association.h"
+#include <iostream>
+
+class PFCandidateRecalibrator : public edm::EDProducer {
+    public:
+        PFCandidateRecalibrator(const edm::ParameterSet&);
+        ~PFCandidateRecalibrator() override {};
+
+    private:
+        void beginRun(const edm::Run& iRun, edm::EventSetup const& iSetup) override;
+        void produce(edm::Event&, const edm::EventSetup&) override;
+
+        edm::EDGetTokenT<std::vector<reco::PFCandidate> > pfcandidates_;
+
+        std::vector<std::tuple<float,float,float>> badChHE_;
+        std::vector<std::tuple<int,int,int,float>> badChHF_;
+};
+
+PFCandidateRecalibrator::PFCandidateRecalibrator(const edm::ParameterSet &iConfig) :
+    pfcandidates_(consumes<std::vector<reco::PFCandidate>>(iConfig.getParameter<edm::InputTag>("pfcandidates")))
+{
+    produces<std::vector<reco::PFCandidate>>();
+    produces<std::vector<reco::PFCandidate>>("discarded");
+    produces<edm::ValueMap<reco::PFCandidateRef>>();
+}
+
+void PFCandidateRecalibrator::beginRun(const edm::Run &iRun, const edm::EventSetup &iSetup)
+{
+    //Get Calib Constants from current GT
+    edm::ESHandle<HcalDbService> GTCond;
+    iSetup.get<HcalDbRecord>().get(GTCond);
+
+    //Get Calib Constants from bugged tag
+    edm::ESHandle<HcalTopology> htopo;
+    iSetup.get<HcalRecNumberingRecord>().get(htopo);
+    const HcalTopology* theHBHETopology = htopo.product();
+    
+    edm::ESHandle<HcalRespCorrs> buggedCond;
+    iSetup.get<HcalRespCorrsRcd>().get("bugged", buggedCond);
+    HcalRespCorrs buggedRespCorrs(*buggedCond.product());
+    buggedRespCorrs.setTopo(theHBHETopology);
+
+    //access calogeometry
+    edm::ESHandle<CaloGeometry> calogeom;
+    iSetup.get<CaloGeometryRecord>().get(calogeom);
+    const CaloGeometry* cgeo = calogeom.product();
+    HcalGeometry* hgeom = (HcalGeometry*)(cgeo->getSubdetectorGeometry(DetId::Hcal,HcalForward));
```

Comment:
is it obvious that all of the conditions accessed above will never change during a run?

Also, if a job is processing a skim and there are multiple runs in the same IOV, it seems wasteful to recompute all values.
Better change to use ESWatchers
https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideHowToGetDataFromES#Seeing_that_a_Record_has_changed